### DISCUSSION TITLE
How to validate GeoJSON

### DISCUSSION THREAD

**Discussion by felixerdy** [colinhacks/zod#2116] [Q&A] [answered]
I'd like to check if an object is a valid [GeoJSON](https://geojson.org/) feature. In the best case, I'd like to limit the feature types (e.g. `Point`, `LineString`, `Polygon`). Is there a smart way to use the already existing GeoJSON namespace?

**Comment by JacobWeisenburger** [comment]
Are you able to give a code example?

**Reply by felixerdy** [reply]
I thought about something like the following. There is `mySchema` where I'd like to validate 3 properties: `title`, `description` and `geometry`. I want to make sure that an incoming geometry object can be validated as one of the GeoJSON Features.

```ts
import * as z from 'zod'

type AllowedGeometry = GeoJSON.Feature<GeoJSON.Point | GeoJSON.LineString | GeoJSON.Polygon>

export const mySchema = z.object({
  title: z.string(),
  description: z.string(),
  geometry: z.object(),    // <-- How to validate AllowedGeometry here?
})
```

**Reply by JacobWeisenburger** [reply]
`Cannot find namespace 'GeoJSON'.`

Please send a reproducible code example.

**Reply by felixerdy** [reply]
Here you can find a stackblitz: https://stackblitz.com/edit/typescript-zpxkoq?file=index.ts
It seems like the [@types/geojson](https://www.npmjs.com/package/@types/geojson) need to be installed

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const mySchema = z.object( {
    title: z.string(),
    description: z.string(),
    geometry: z.object( {
        type: z.literal( 'Feature' ),
        properties: z.object( {} ),
        geometry: z.object( {
            coordinates: z.number().array(),
            type: z.literal( 'Point' ),
        } ),
    } ),
} )

console.log(
    mySchema.safeParse( {
        title: 'ABC',
        description: 'DEF',
        geometry: {
            type: 'Feature',
            properties: {},
            geometry: {
                coordinates: [ 20.393936829428668, 17.253197020069166 ],
                type: 'Point',
            },
        },
    } ).success
) // true
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by felixerdy** [reply]
That's not exactly what I'm looking for, I thought maybe it's possible to infer the schema from the @types/geojson types

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
type AllowedGeometry = GeoJSON.Feature<
    GeoJSON.Point | GeoJSON.LineString | GeoJSON.Polygon
>

const geometrySchema = z.object( {
    type: z.literal( 'Feature' ),
    properties: z.object( {} ),
    geometry: z.object( {
        coordinates: z.number().array(),
        type: z.literal( 'Point' ),
    } ),
} ) satisfies z.ZodType<AllowedGeometry>
```

**Comment by chiubaca** [comment]
I was thinking it would be useful if there was a `z.geojson()`. cc @colinhacks would you accept a PR for this?

**Reply by el-pendeloco** [reply]
In the meantime, I've cooked up a a library with that feature, because I also needed it: https://www.npmjs.com/package/@infra-blocks/zod-utils. [Source code](https://github.com/infrastructure-blocks/ts-zod-utils)

**Reply by chiubaca** [reply]
Nice!!
